python/cuml/cuml/accel/accelerator.py

```diff
@@ -0,0 +1,335 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import annotations
+
+import importlib
+import importlib.abc
+import importlib.machinery
+import importlib.util
+import sys
+import types
+from collections.abc import Iterable
+from threading import RLock
+from typing import Any, Callable, Union
+
+__all__ = ("Accelerator",)
+
+
+PatchType = Union[
+    Callable[[types.ModuleType], dict[str, Any]],
+    dict[str, Any],
+    str,
+]
+
+
+class AccelModule(types.ModuleType):
+    """An accelerated module.
+
+    Wraps an underlying module, replacing access to certain attributes
+    with accelerated versions.
+    """
+
+    # Internal state. All have prefix `_accel_` to try and avoid accidental collisions
+    # with names on the wrapped module.
+    _accel_module: types.ModuleType
+    _accel_denylist: frozenset[str]
+    _accel_overrides: dict[str, Any]
+
+    def __repr__(self) -> str:
+        return super().__repr__().replace("<module", "<accelerated module", 1)
+
+    def __getattr__(self, name: str) -> Any:
+        if (accelerated := self._accel_overrides.get(name)) is not None:
+            # This has an override and could be accelerated. We first need
+            # to check that the accessing module isn't in the denylist.
+            #
+            # To do this we walk up the stack, skipping frames in importlib
+            frame = sys._getframe()
+            while True:
+                assert frame.f_back is not None
+                # Get the module name of the caller (if available)
+                modname = frame.f_back.f_globals.get("__name__")
+                if modname is not None:
+                    modname = modname.split(".", 1)[0]
+                    if modname == "importlib":
+                        # Caller is in importlib, continue up the stack
+                        frame = frame.f_back
+                        continue
+                # Found a valid module name
+                break
+
+            if modname not in self._accel_denylist:
+                # Not in denylist, use the accelerated version
+                return accelerated
+
+        return getattr(self._accel_module, name)
+
+    def __setattr__(self, name: str, val: Any) -> None:
+        if name in ("_accel_module", "_accel_denylist", "_accel_overrides"):
+            super().__setattr__(name, val)
+        else:
+            setattr(self._accel_module, name, val)
+
+    def __dir__(self) -> list[str]:
+        return dir(self._accel_module)
+
+
+def wrap_module(
+    module: types.ModuleType, denylist: Iterable[str] | None = None
+) -> AccelModule:
+    """Create a new AccelModule instance.
+
+    Parameters
+    ----------
+    module : ModuleType
+        The original, unaccelerated module.
+    denylist : sequence, optional
```

Comment:
This denylist is pretty clear to folks that have worked in these areas, but it might be good for mantainability in the medium and longer term to add an example of it being used (like why is `sklearn` and `treelite` on denylist so that they work well together) so that it becomes easy and intuitive for an entirely newcomer to the codebase.

Reply:
After #6565 (moving this to a function w/ some comments) this should hopefully be clearer.